### ISSUE TITLE
Feature Request: Like monsters roll initiative together

### ISSUE THREAD

**Issue by malacandrian** [valentine195/obsidian-initiative-tracker#17]
# Request
As a dungeon master; I want all monsters that use the same stat block to [act on the same initiative](https://www.5esrd.com/gamemastering/combat/); so that combat has less context switching, making it more predictable and smoother for the players. 

Current behaviour: every monster rolls initiative independently.

Desired behaviour: Having an option when defining an encounter to group monsters that will act on the same turn. It should allow this without interrupting the current behaviour, or requiring aditional syntax for dungeon masters who want to continue using the current behaviour.

# Possible implementations
## Global flag - group by stat block
Having a settings flag that universially switches the behaviour on or off. It groups monsters by comparing the origin of each monster (be it SRD Monster, Homebrew, or inline stats).

eg, for the following encounter, all the goblins will be grouped, but Inline A and Inline B won't

```encounter
name: Example 1
creatures:
    - 1: Goblin
    - 1: Inline A, 7, 15, 2
    - 1: Goblin
    - 1: Inline B, 7, 15, 2 
    - 5: Goblin
```

### Advantages:
- It's the most literal way of interpreting the SRD's rule about like monsters acting together.
- It feels simple at first glance
- There's no way of it interfering with current default behaviour

Disadvantages
- It's very inflexible
- From a quick look through the code, I don't think the monsters currently store their origin, though I might be mistaken
- Global solution - no ability to decide on an encounter-by-encounter basis whether or not to use it
    - This could be solved by adding an optional aditional flag to the encounter block, with the global flag controlling the default for if the optional flag is not provided
- There are edge cases where the "correct" behaviour is not obvious. e.g.
    ```encounter
    name: Example 2
    creatures:
        - 1: Inline, 7, 15, 2
        - 1: Inline, 7, 15, 2
        - 1: Inline, 10,15,2
        - 1: Inline, 7,15,3
    ```
    - How many groups of monsters should this create?
        -  4? Are differently declared Inline creates different?
        - 1? They all share the same name, does that make them the same? What initiative bonus should the group have - 2 or 3? Somewhere inbetween?
        - 3? One per unique stat block per name? What if I roll individual monster HP but still want them grouped, should it be 2?

## Global flag - group by line.
Use the same setting flag as above, but this time creating one group per line. So Example 1 from above would create 5 groups under this system, whereas it created 3 earlier.

### Advantages:
- Answers the awkward egde cases from Example 2
- Still relatively simple
- Adds some flexibility

### Disadvantages
- Doesn't allow grouping of disimilar monsters
   - On the one hand, this could be feature creep, after all the SRD does only say "The GM makes one roll for an entire group of **identical** creatures"
   - On the other hand, we have the question from before about rolling individual HP for like monsters, are they still identical? It feels like 5 goblins with different HP should still be able to act on the same initiative order.

## New syntax for groups
Instead of using any kind of flags, aditional syntax is added to build monster groups.

e.g. (_Syntax as example only, will need futher work._)
```encounter
name: Example 3
creatures:
    - 1: Hobgoblin Chieftan
    - 
        name: Melee Minions
        mod: 2
        creatures:
            - 2: Goblin Bandit
            - 2: Goblin Thug
    - 
        name: Artillerist Minions
        creatures:
             - 4: Goblin Archer
```

In this example there are 3 groups: the boss, and two groups on minions. The name and mod keys are optional. If the mod is excluded, it takes the average mod of all the monsters in the group.

### Advantages
- Maximum flexibility
- No global settings
- Precise enough to let the dungeon master specify exactly what they want, avoiding the edge cases
- Because monsters can still exist at the top level, it shouldn't interfere with current behaviour

### Disadvantages
- Very complex
- The main use case is for identical creatures (like the Goblin Archers in example 3) this now requires a lot more ceremony.
    - Perhaps there is some additional syntax to add for this, e.g. `G4: Goblin archer` indicates 4 Goblin archers who will act as a group?
    - Perhaps this could be combined with Global flag - group by line. to solve it

**Comment by seansbox** [comment]
Awesome request. Any of these approaches would cover my use cases as well.

**Comment by valentine195** [comment]
The plugin should already be grouping creatures in encounter blocks if their stats are actually the same. Is it not?

It shouldn't be too hard to enable this in the initiative tracker side as well, once I have some time to work on it!

**Comment by seansbox** [comment]
My experience has been a 3: Goblin line will result in three different goblin initiative rolls/lines in the tracker.

**Comment by valentine195** [comment]
Yep - it is definitely not grouping enemies in the tracker. I just read the original FR as saying it also wasn't doing it in the encounter code blocks, which it *should be*!

**Comment by malacandrian** [comment]
> Yep - it is definitely not grouping enemies in the tracker. I just read the original FR as saying it also wasn't doing it in the encounter code blocks, which it _should be_!

To confirm: it does succesfully group on the encounter block, just not on the initiative roll.

**Comment by valentine195** [comment]
Okay good!

like I said, I already have the logic for grouping same-stat creatures. Shouldn’t be too hard to move it into the tracker as well once I have some time!

itll likely be a global setting as well as a per-encounter flag / a button in the tracker view to group them.

**Comment by malacandrian** [comment]
Brilliant, thanks for being on this.

**Comment by dschibster** [comment]
Just chiming in to say that this would also save my day. Hoping it can get pushed live soon. :))

**Comment by valentine195** [comment]
I actually worked a bit on this today. Am going to roll this in with a couple other changes, hopefully tomorrow.

**Comment by valentine195** [comment]
Just wanted to let you guys know that I did some backend work (and work on the Encounter statblock) today that will help with this feature. Coming soon, I promise!

**Comment by valentine195** [comment]
This should be enabled in 5.0.2 under the Group Creatures option in the initiative tracker view menu as well as a global setting in settings.
